File: nim/src/coding_practice/chapter1_2.nim
# Write code to reverse a C-style string. (C-string means that "abcd" is
# represented as five characters, including the null character.)

proc rev*(str: var string) =
  ## nim solution; does not handle c-strings
  let last = len(str) - 1
  for idx, _ in str:
    if idx > last - idx: break
    (str[idx], str[last-idx]) = (str[last-idx], str[idx])

File: elixir/lib/chapter1_2.ex
defmodule Crack.Chapter1_2 do
  @moduledoc """
  1.2 C-String Rerversal:

  Write code to reverse a C-Style string. (C-String means that "abcd" is
  represented as five characters, including the null character.)
  """

  @doc """
  Reverse a C-Style string.

      iex> Crack.Chapter1_2.start("")
      ""

      iex> Crack.Chapter1_2.start("abba")
      "abba"

      iex> Crack.Chapter1_2.start("abcd")
      "dcba"

      iex> Crack.Chapter1_2.start("hello world!")
      "!dlrow olleh"

  """
  def start(""), do: ""

  def start(str) when is_binary(str), do: reverse(str)

  defp reverse(<<>>), do: <<>>

  defp reverse(<<head::utf8, tail::binary>>) do
    <<reverse(tail)::binary, head::utf8>>
  end
end
